### ISSUE TITLE
Need compression decoder code

### ISSUE THREAD

**Issue by CassieAinsworthx** [noamelata/AdaSense#1]
Dear Author :
       May i ask when the code for compression will be released? And i have a question about the work of compression, in the encoder i calculate for y_0 ,and quantize it for transmit , and in the decoder , we know y_0 , but we do not know H, then how the decoder continue to work? In the sample process need to use H_pinv, this process need H， I must be misunderstood the paper ,but i just can't think through it . Thank you,  looking forward to your reply.
                                                                                                                               Sincerely, k.

**Comment by noamelata** [comment]
Hi,

We have another paper under review dedicated to image compression with AdaSense, you can find it [here](https://arxiv.org/abs/2407.09896). In this paper we answer your question, and show how using AdaSense for both the encoder and decoder enables the selection of a matching $H$ on both sides, as long as they share a random seed. As you wrote, it is not trivial from the original fomulation, so I advise reading the manuscript.
The full code release may take a bit more time. In the meantime, you can use the code in this repository as an implementation of AdaSense with DDRM as the posterior sampler, and follow the pseudo-code below:
```python
from utilities import posterior_sampler, restoration_fuction, entropy_encode, entropy_decode, set_random_seed

def AdaSense_Step(H, y, r, shape, s=None):
    c, h, w = shape
    s = s if s is not None else (r * 4) // 3
    noise = torch.randn((s, c, h, w))
    samples = posterior_sampler(noise, H, y)
    samples = samples.reshape(s, -1)
    samples = samples - samples.mean(0, keepdim=True)
    new_rows = torch.linalg.svd(samples, full_matrices=False)[-1][:r]
    return new_rows

def PSC_compress(image, N, r, seed=0):
    set_random_seed(seed)
    c, h, w = image.shape
    H = torch.zeros((0, c * h * w))         # Empty sensing matrix
    y = H @ image.reshape((-1, 1))          # Empty measurements
    compressed_representation = y.clone()
    
    for n in range(N):
        new_rows = AdaSense_Step(H, y, r, (c, h, w)) 
        H = torch.cat([H, new_rows])
        y = torch.cat([y, new_rows @ image.reshape((-1, 1)])
        
        compressed_representation = y.to(torch.float8_e4m3fn) # Quantize
        y = compressed_representation.to(torch.float32)
        
    return entropy_encode(compressed_representation)

def PSC_decompress(compressed_representation, N, r, seed=0):
    set_random_seed(seed)
    compressed_representation = entropy_decode(compressed_representation)
    c, h, w = image.shape
    H = torch.zeros((0, c * h * w))         # Empty sensing matrix
    y = H @ image.reshape((-1, 1))          # Empty measurements
    
    for n in range(N):
        new_rows = AdaSense_Step(H, y, r, (c, h, w))
        H = torch.cat([H, new_rows])
        y = compressed_representation[:(n*r)].to(torch.float32)
        
    return restoration_fuction(H, y)
```

Best,
Noam

**Comment by CassieAinsworthx** [comment]
Dear Author,
        Thank you so much for your reply.  I've already read the paper you refered weeks ago, and it's very inspiring to me. I was in doubts about how to use y on decoding adasense step, and now i understood after seeing your post code, thank you so much, and wish you all the best.
Sincerely, kexin.
